ARCHITECTURE.md

```diff
@@ -0,0 +1,339 @@
+object as the Resource they are linked from. This enables that at some point, a
+function like `saveSolidDatasetAt` might update not just the data in the
+Resource itself, but also that in one or more linked Resources.
+
+This concern was mainly brought up by @pmcb55, so further questions regarding
+this can be directed to him.
+
+# Concepts
+
+## Resources, Files and SolidDatasets
+
+solid-client refers to everything that can be fetched from a URL as a Resource.
+There are two types of Resources: those containing RDF in a serialisation that
+the server understands ("SolidDatasets"), and those that do not ("Files").
+
+The main difference between the two is that, by virtue of the server
+understanding the structure of the former, we can do partial updates (i.e.
+PATCH requests) and retrieve the data in different formats (currently, the
+spec mandates that servers can serve such data as at least Turtle and JSON-LD).
+
+Thus, data stored as Turtle or JSON-LD can be fetched as a SolidDataset; the
+rest are just treated as regular Files: while might be parsed and manipulated by
+other libraries, solid-client just allows downloads and uploads. This applies to
+binary file types like JPEG, WebM and `.txt`, but also to files containing
+structured data in a non-RDF format, like JSON, XML or OpenDocument, or even
+to files containing structured data in an RDF format not understood by the
+server, such as RDFa or RDF-XML.
+
+## `With*` types
+
+Although SolidDatasets and Files (see previous section) are different types of
+Resources, there is certain data that is common to all Resources. This data is
+set on properties for either type of Resource, which is specified in a `With*`
+type.
+
+For example, there is some metadata determining where a particular Resource is
+fetched from; this is set on the property `internal_resourceInfo` for both
+SolidDatasets and Files, as specified by the `WithResourceInfo` type. Whether
+a SolidDataset or a File has that metadata available is then available using a
+Union Type, e.g. `SolidDataset & WithResourceInfo`.
+
+This means that we can define a function like `getContentType` that doesn't
+really care whether its argument is a SolidDataset or a File, as long as its
+parameter has that metadata available:
+
+function getContentType(resource: WithResourceInfo)
+
+Note that in our API documentation, we explicitly reserve the right to change
+the implementation of these data structures. If developers want to access the
+data contained in them, they should use the functions we provide for them, such
+as `getContentType`. We can, of course, use those internally as well.
+
+# Module map
+
+## `src/resource/*`
+
+Everything related to fetching and storing data. `file.ts` has everything
+related to fetching Files, which is mostly a thin wrapper around a regular
+`fetch`, adding some metadata. `solidDataset.ts` is similar, but attempts to
+parse the fetched data as RDF in order to turn it into a SolidDataset. And
```

Comment:
Maybe mention that how the parsing is specified is described in the `src/format` section ?

Reply:
80121b8e